File: bfsdfs.py
# BFS Implementation
def bfs(graph, start):
    queue = [start]  # Use a list as a queue
    visited = set([start])
    order = []

    while queue:
        node = queue.pop(0)  # Remove from the front (FIFO)
        order.append(node)

        for neighbor in graph[node]:
            if neighbor not in visited:
                visited.add(neighbor)
                queue.append(neighbor)

    return order

# DFS Implementation
def dfs(graph, start):
    stack = [start]  # Use a list as a stack
    visited = set([start])
    order = []

    while stack:
        node = stack.pop()  # Remove from the top (LIFO)
        order.append(node)

        for neighbor in reversed(graph[node]):  # Reverse to maintain left-to-right order
            if neighbor not in visited:
                visited.add(neighbor)
                stack.append(neighbor)

    return order

# Function to convert numbers to letters
def number_to_letter(order):
    return [chr(96 + num) for num in order]

# Example Usage
graph = {
    1: [2, 3],
    2: [1, 4, 5],
    3: [1],
    4: [2, 6],
    5: [2],
    6: [4]
}

start_node = 1

# BFS Traversal
bfs_order = bfs(graph, start_node)
bfs_letters = number_to_letter(bfs_order)
print("BFS Order (Letters):", bfs_letters)

# DFS Traversal
dfs_order = dfs(graph, start_node)
dfs_letters = number_to_letter(dfs_order)
print("DFS Order (Letters):", dfs_letters)

File: bfsdfs.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <stack>
#include <unordered_set>

using namespace std;

// BFS Implementation
vector<int> bfs(const vector<vector<int>>& graph, int start) {
    queue<int> q;
    unordered_set<int> visited;
    vector<int> order;

    q.push(start);
    visited.insert(start);

    while (!q.empty()) {
        int node = q.front();
        q.pop();
        order.push_back(node);

        for (int neighbor : graph[node]) {
            if (visited.find(neighbor) == visited.end()) {
                visited.insert(neighbor);
                q.push(neighbor);
            }
        }
    }

    return order;
}

// DFS Implementation
vector<int> dfs(const vector<vector<int>>& graph, int start) {
    stack<int> s;
    unordered_set<int> visited;
    vector<int> order;

    s.push(start);
    visited.insert(start);

    while (!s.empty()) {
        int node = s.top();
        s.pop();
        order.push_back(node);

        // Traverse neighbors in reverse order to maintain left-to-right order
        for (auto it = graph[node].rbegin(); it != graph[node].rend(); ++it) {
            int neighbor = *it;
            if (visited.find(neighbor) == visited.end()) {
                visited.insert(neighbor);
                s.push(neighbor);
            }
        }
    }

    return order;
}

int main() {
    // Graph representation as an adjacency list
    vector<vector<int>> graph = {
        {},         // 0 (unused)
        {2, 3},     // 1
        {1, 4, 5},  // 2
        {1},        // 3
        {2},        // 4
        {2}         // 5
    };

    int start_node = 1;

    // BFS Traversal
    vector<int> bfs_order = bfs(graph, start_node);
    cout << "BFS Order: ";
    for (int node : bfs_order) {
        cout << node << " ";
    }
    cout << endl;

    // DFS Traversal
    vector<int> dfs_order = dfs(graph, start_node);
    cout << "DFS Order: ";
    for (int node : dfs_order) {
        cout << node << " ";
    }
    cout << endl;

    return 0;
}